File: examples/estimate_fee.rs
//! Print the current priority fee distribution for a given RPC.
//!
//! Usage:
//!     cargo run --example estimate_fee -- <RPC_URL>
//!     cargo run --example estimate_fee -- https://api.mainnet-beta.solana.com

use std::str::FromStr;
use std::sync::Arc;

use solana_client::nonblocking::rpc_client::RpcClient;
use solana_landed_tx::FeeEstimator;
use solana_sdk::pubkey::Pubkey;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rpc_url = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "https://api.mainnet-beta.solana.com".to_string());

    // Optional 2nd arg: a writable account to scope the fee query to slots
    // that touched it. Defaults to the USDC mint, which is touched constantly.
    let account = std::env::args()
        .nth(2)
        .unwrap_or_else(|| "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v".to_string());
    let writable = Pubkey::from_str(&account)?;

    println!("RPC:     {rpc_url}");
    println!("Scoped to writable account: {writable}");

    let rpc = Arc::new(RpcClient::new(rpc_url));
    let estimator = FeeEstimator::new(rpc);

    let est = estimator.estimate(&[writable]).await?;

    println!("samples: {}", est.samples);
    println!("p50:     {} µLamports/CU", est.p50);
    println!("p75:     {} µLamports/CU", est.p75);
    println!("p90:     {} µLamports/CU", est.p90);
    println!("p95:     {} µLamports/CU", est.p95);
    println!("max:     {} µLamports/CU", est.max);
    println!("mean:    {} µLamports/CU", est.mean);

    let cu_limit: u64 = 200_000;
    let lamports = est.p75.saturating_mul(cu_limit) / 1_000_000;
    println!("\nAt p75 with a {cu_limit} CU budget, you'd pay {lamports} lamports in priority fees.");

    Ok(())
}

File: python/examples/estimate_fee.py
"""Print the current priority fee distribution for a given Solana RPC.

Usage:
    python examples/estimate_fee.py
    python examples/estimate_fee.py https://api.mainnet-beta.solana.com
    python examples/estimate_fee.py https://api.mainnet-beta.solana.com <PUBKEY>
"""

import sys

from solana_landed_tx import FeeEstimator

DEFAULT_RPC = "https://api.mainnet-beta.solana.com"
USDC_MINT = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"


def main() -> None:
    rpc_url = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_RPC
    writable = sys.argv[2] if len(sys.argv) > 2 else USDC_MINT

    print(f"RPC:     {rpc_url}")
    print(f"Scoped to writable account: {writable}")

    est = FeeEstimator(rpc_url)
    result = est.estimate([writable])

    print(f"samples: {result.samples}")
    print(f"p50:     {result.p50} µLamports/CU")
    print(f"p75:     {result.p75} µLamports/CU")
    print(f"p90:     {result.p90} µLamports/CU")
    print(f"p95:     {result.p95} µLamports/CU")
    print(f"max:     {result.max} µLamports/CU")
    print(f"mean:    {result.mean} µLamports/CU")

    cu_limit = 200_000
    lamports = (result.p75 * cu_limit) // 1_000_000
    print(
        f"\nAt p75 with a {cu_limit} CU budget, "
        f"you'd pay {lamports} lamports in priority fees."
    )


if __name__ == "__main__":
    main()
